crates/fluvio-connector-package/src/renderer.rs

```diff
@@ -0,0 +1,174 @@
+        let secret_store = crate::secret::default_secret_store()?;
+        let mut env = Environment::new();
+        env.set_syntax(minijinja::Syntax {
+            block_start: "${%".into(),
+            block_end: "%}".into(),
+            variable_start: "${{".into(),
+            variable_end: "}}".into(),
+            comment_start: "${*".into(),
+            comment_end: "*}".into(),
+        })?;
+
+        Ok(Self::new(secret_store, env))
+    }
+
+    pub fn render_value(&self, input: &serde_yaml::Value) -> anyhow::Result<serde_yaml::Value> {
+        match input {
+            Value::String(s) => {
+                let resolved = self.render_str(s)?;
+                Ok(Value::String(resolved))
+            }
+            Value::Mapping(mapping) => {
+                let mut result = Mapping::new();
+                for (key, value) in mapping.iter() {
+                    let resolved = self.render_value(value)?;
+                    result.insert(key.clone(), resolved);
+                }
+                Ok(Value::Mapping(result))
+            }
+            Value::Sequence(seq) => {
+                let mut result = Sequence::new();
+                for value in seq.iter() {
+                    let resolved = self.render_value(value)?;
+                    result.push(resolved);
+                }
+                Ok(Value::Sequence(result))
+            }
+            other => {
+                // ignore numbers, booleans and null for now
+                Ok(other.clone())
+            }
+        }
+    }
+
+    fn render_str(&self, input: &str) -> anyhow::Result<String> {
+        let context = self.build_context(input)?;
+        match self.env.render_str(input, context) {
+            Ok(rendered) => Ok(rendered),
+
+            Err(e) => Err(anyhow::anyhow!("failed to render: `{}`. {}", input, e)),
+        }
+    }
+
+    fn build_context(&self, input: &str) -> anyhow::Result<Context> {
+        let secrets: HashMap<_, _> = detect_secrets_from_str(input)?
+            .into_iter()
+            .map(|s| {
+                self.secret_store
+                    .read(&s)
+                    .map(|secret_value| (s, secret_value))
+            })
+            .collect::<Result<HashMap<_, _>, _>>()?;
+        Ok(Context { secrets })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::io::Write;
+
+    use crate::secret::{self, FileSecretStore};
+
+    use super::ConfigRenderer;
+
+    #[test]
+    fn test_build_context() {
+        let mut file = tempfile::NamedTempFile::new().expect("failed to create tmp file");
+        file.write_all(b"foo=bar\napi_key=my_api_key\ninterval=10s")
+            .expect("file to write");
+        secret::set_default_secret_store(::std::boxed::Box::new(FileSecretStore::from(
+            file.path(),
+        )))
+        .expect("failed to set default secret store");
+
+        let renderer = ConfigRenderer::new_with_default_stores().unwrap();
+        let context: crate::context::Context =
```

Comment:
Why do you need ConfigRenderer here?

Reply:
because `build_context` is a method of `ConfigRenderer` that uses internal secret_store to build the context